### ISSUE TITLE
Support for electron/fuses

### ISSUE THREAD

**Issue by eXhumer** [castlabs/electron-releases#152]
Is there any consideration for supporting [electron/fuses](https://github.com/electron/fuses)? I would like to encrypt the cookies with `[FuseV1Options.EnableCookieEncryption]: true` fuse with my application, but making any fuse changes causes the `python -m castlabs_evs.vmp sign-pkg` command to fail with the following error. It makes binary level changes to executable to flip fuses, which causes the bad signature I think.

```
(evs-env) PS C:\Users\eXhumer\Projects\eXViewer> python -m castlabs_evs.vmp sign-pkg .\out\eXViewer-win32-x64\
Signing: .\out\eXViewer-win32-x64\eXViewer.exe
 - Verifying existing VMP signature
 - Signature invalid: Failed to verify signature
 - Requesting VMP signature
HTTPError: Signing request failed: 400 Bad Request
 <- ValidityError: Binary signature denied
```

The `sign-pkg` command has no issues if no fuses are changed from default.

By default, the following fuses are used

```
(evs-env) PS C:\Users\eXhumer\Projects\eXViewer> yarn run electron-fuses read --app .\out\eXViewer-win32-x64\eXViewer.exe      yarn run v1.22.19
$ C:\Users\eXhumer\Projects\eXViewer\node_modules\.bin\electron-fuses read --app .\out\eXViewer-win32-x64\eXViewer.exe
Analyzing app: eXViewer.exe
Fuse Version: v1
  RunAsNode is Enabled
  EnableCookieEncryption is Disabled
  EnableNodeOptionsEnvironmentVariable is Enabled
  EnableNodeCliInspectArguments is Enabled
  EnableEmbeddedAsarIntegrityValidation is Disabled
  OnlyLoadAppFromAsar is Disabled
  LoadBrowserProcessSpecificV8Snapshot is Disabled
Done in 0.49s.
```

**Comment by khwaaj** [comment]
This is actually something that would be interesting to add support for. For now it is not planned, and it is likely to be rather complicated to add safely (so I can't give an estimate or anything), but I will make sure it is put up for consideration.

**Comment by Kilian** [comment]
I would be good to at this to the FAQ because, as of 2025, it seems to still be an issue. A lot of the options you'd set with fuses make Electron _more_ secure, so its actually hardening the app.

**Comment by khwaaj** [comment]
Yes, I agree this should be in the FAQ - I'll make sure it gets added.

For now the only way to be able to use fuses with ECS is to go the 3PL certification route, which would allow setting your own fuses (as well as making other changes). This route comes associated with a cost though.

I do see the value of this being available for free users which do not have the ability/resources to go the 3PL route though, so we are not ruling out supporting this at some point in the future. Unfortunately the required updates to support this are not trivial and would require some investment on our part, so it has not been planned at this time.

**Comment by khwaaj** [comment]
@Kilian, thanks for the suggestion - using a set of static fuses for might be a viable alternative. I'm also part of the e-mail exchange, but since this is a public matter it makes more sense to keep that part of the discussion here.

I suppose some of these defaults may have consequences for some use-cases,  e.g. during local testing. Do you see any other potential drawbacks of doing it this way?

**Comment by Kilian** [comment]
> I suppose some of these defaults may have consequences for some use-cases, e.g. during local testing. Do you see any other potential drawbacks of doing it this way?

@khwaaj yes, some of these, like `OnlyLoadAppFromAsar` and `RunAsNode` would interfere with local development (fuses are mostly settings that the electron team would like to have as defaults but break too much things for dev)

Pragmatically, shipping an electron-development (without fuses) and an electron-releases (with fuses) binary would be a good solution for this though I understand that that also doubles the amount of work for each release. Alternatively, having people use the regular electron release during development and then build using castlabs/electron-releases would probably also be workable for most.

**Comment by khwaaj** [comment]
We are looking at supporting a static set of fuses in the near future. This is in addition to no fuses, which will still be what is in the ECS release binaries, meaning we will not break any existing use of ECS (i.e. we will support signing the default set of fuses as well as one additional set we settle on)..

As a starting point we are considering the fuses suggested by @Kilian above, but is the `IS_WIN` conditional really necessary? The documentation suggests that ASAR Integrity is supported on Windows for versions 30 and later, and with version 32 being the oldest supported version we are past that requirement. Or is there some other reason ASAR Integrity is not a good default on Windows?

**Comment by Kilian** [comment]
@khwaaj It seems like that's just a leftover from when I initially set up fuses! It indeed works fine in current versions of Electron. I've updated my example above to reflect that.

**Comment by Kilian** [comment]
That sound amazing, can't wait!

**Comment by khwaaj** [comment]
All releases starting with `v35.0.0` should now support signing with the above set of fuses. I'll see about updating the FAQ to reflect this update as well.

**Comment by eXhumer** [comment]
Hey. Thanks for adding the custom fuses support. You should probably document the set of fuses needed to be set to be a valid binary signature for signing. Other than that, it works great from my testing. I'm closing this as completed.
